extensions/amp-video/0.1/video-cache.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {createElementWithAttributes} from '../../../src/dom';
+import {toArray} from '../../../src/types';
+
+/** @private {?boolean} whether the document can use cache url */
+let docSupportsCache = null;
+
+/**
+ * Add the caching sources to the video if opted in.
+ * @param {!AmpVideo} video
+ * @return {!Promise}
+ */
+export function addCacheSources(video) {
+  if (!triggerVideoCache(video.win)) {
+    return Promise.resolve();
+  }
+  const servicePromise = Services.cacheUrlServicePromiseForDoc(video.element);
+  return servicePromise
+    .then((service) =>
+      Promise.all([
+        service.createCacheUrl(selectVideoSource(video.element)),
+        Services.storyRequestServiceForOrNull(video.win),
+      ])
+    )
+    .then((promiseResult) => {
+      const requestUrl = promiseResult[0].replace('/c/', '/mbv/');
+      return promiseResult[1].executeRequest(requestUrl);
+    })
+    .then((response) => applySourcesToVideo(video.element, response['sources']))
+    .catch(() => {});
```

Comment:
Why might this promise chain reject? Should we at least log an error message?

Reply:
The GET request to the AMP CDN might fail (if the backend is not caching these videos or the video source is not in the cdn), which should not break amp-video so we need to catch the error. Note that a request failing will show an error in the console already, so no need to duplicate an error.

Reply:
Suggestion: Include above explanation (at least in part) as a comment